Sizebay - Front-end Challenge

guusdias opened the pull request:
Hey there!

This is my application for the Junior Developer job role at Sizebay.
Feel free to clone and run it on your machine with the following commands:

1. First, install dependencies with:
```
yarn
```

2. Then, run the application with:
```
yarn wrap
```

Hope you enjoy it!

notlesz reviewed (changes requested):
Olá [@Gustavo](https://github.com/guusdias), obrigado por enviar o desafio e aqui está o review. Sua entrega foi bem consistente, encontrei alguns erros que foram apontados e também deixei pontos de melhoria tanto no layout e código

# Layout 

### Erros:

- Não permite excluir ou retomar a tarefa para pendente quando está concluída. O comportamento ideal é deixar sempre editável.
- O tooltip não está se comportando como deveria após clicar em uma tarefa.
- Ao completar todas as tarefas e criar uma nova, as tarefas anteriores são limpas. O comportamento ideial é persistir todos os dados anteriores ao criar uma nova tarefa.

### Melhorias:

- Adicione feedback visual se uma tarefa estiver concluída: Exemplos: ![enter image description here](https://private-user-images.githubusercontent.com/79087985/306651946-b5e75b89-7dcf-411d-8b46-1c7ac1d61392.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTAzMDA0NzUsIm5iZiI6MTcxMDMwMDE3NSwicGF0aCI6Ii83OTA4Nzk4NS8zMDY2NTE5NDYtYjVlNzViODktN2RjZi00MTFkLThiNDYtMWM3YWMxZDYxMzkyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMTMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzEzVDAzMjI1NVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTM2ODdhOWZiNmFhYTJmMDg2OTdiZGEzMDBhNTNjYzU5N2MwOTU4OGIwMjEyZTdkYTA1MWI3NGI2ZDMwMzVmYjYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.hNCHSOOEr7XVOHMqFM2MnBgJEGjgdZCoetcQTXC88x0)
- Desative o botão e a submissão do input para criar uma tarefa e remova o comportamento de opacidade.

# Code

- Utilize a criação de temas do styled-components para definir os estilos necessários e distribuí-los em sua aplicação; dessa forma, você não precisa criar variáveis no CSS.
- Melhore a nomenclatura de seus `states`, `const` e `functions` para uma melhor DX (Developer Experience).
- Utilize as novas features do [React](https://react.dev/) que melhoram drasticamente o desempenho de sua aplicação, por exemplo: `useMemo`, `useCallback`, `useDeferredValue`. Estude e entenda em quais momentos são necessários utilizá-las.

---


Aguardaremos seus ajustes pós-revisão para prosseguir com o processo. Qualquer dúvida estou a disposição :)

notlesz commented on src/components/TasksList/index.tsx:
O retorno desse componente está um pouco verboso, que tal melhorar ele?

Dica: transforme os alertas em componentes menores e controle a sua renderizaçao com base na regra estabelecida pra cada.

** Dica extra: o filtro de tarefas está correto mas você pode diminuir ainda mais sua complexidade.

notlesz commented on src/components/TaskButtons/index.tsx:
Esse arquivo está exportando um elemento estilizado e não um componente React. Transforme-o em componente ou deixei como estilo onde está sendo utilizado.

guusdias commented:
# Challenge Corrections :wrench:

First of all, I want to thank you so much, @notlesz, for your feedback. I really appreciate your quick response and thoughts on the project.

## Layout :nail_care: 

To address the issue of returning a task once it's completed, I first adjusted the visibility of the buttons. Then, I introduced a new type of button specifically for returning tasks. This button is colored orange, which is not part of the design system. The functionality to clear all tasks has been removed.

For completed tasks, I've applied a line-through text decoration to indicate to the user that the task has been finished.

I've also removed the unnecessary opacity behavior from the input bar. Additionally, the `ThemeProvider` was set up to assist with styling the components.

___

## Code :abacus: 

To improve readability and organization, I separated the alerts into a different component, resulting in a cleaner listing component. I also revised a ternary operator that was previously confusing and nearly incomprehensible.

The unused button component was removed, and I opted to create it directly within the specific parent component.

For now I decided not to apply those hooks you mentioned, but for sure very soon I'll be learning how to apply them correctly.

---

Hope you guys enjoy my corrections! If there are any further suggestions or feedback, please feel free to share. We're always open to improvements ;)

Thank you for the effort and dedication from everyone!

notlesz commented:
Olá @guusdias, beleza? Fiz o review dos seus ajustes e gostei bastante das melhorias tanto no código e layout. O plus da entrega seria a utilização dos hooks do React, o que mostra que você entende o funcionamento de cada um e quando usar, mas entendo o motivo de não ter utilizado e valeu por explicar 😄.

Obrigado por dedicar seu tempo na entrega desse desafio, entraremos em contato com as próximas etapas 🚀

vaporwavie commented:
Congrats on your submission @guusdias! Accepting the challenge and closing this now.
